Sources/StreamChat/StateLayer/ChatListState.swift

```diff
@@ -0,0 +1,53 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// Represents a list of channels matching to the specified query.
+@available(iOS 13.0, *)
+public final class ChatListState: ObservableObject {
+    private let observer: Observer
+    private let clientConfig: ChatClientConfig
+    let query: ChannelListQuery
+    
+    init(channels: [ChatChannel], query: ChannelListQuery, clientConfig: ChatClientConfig, database: DatabaseContainer) {
+        self.channels = channels
+        self.clientConfig = clientConfig
+        hasLoadedAllChannels = channels.count < query.pagination.pageSize
+        observer = Observer(query: query, chatClientConfig: clientConfig, database: database)
+        self.query = query
+        
+        observer.start(
+            with: .init(channelsDidChange: { [weak self] changes in
+                guard let self else { return }
+                await self.setValue(self.orderedChannels.withListChanges(changes), for: \.channels)
+            })
+        )
+    }
+    
+    /// An array of channels for the specified ``ChannelListQuery``.
+    @Published public private(set) var channels = [ChatChannel]()
+    
+    /// True, if all the channels were loaded with paginated loading.
+    @Published public private(set) var hasLoadedAllChannels = false
+    
+    // MARK: - Mutating the State
```

Comment:
Since this is ObservableState then here are MainActor annotated functions which will force mutations on the main thread and therefore, should cause less issues with threading and using it in SwiftUI views directly.

But this is not safe from cases where someone reads properties from background threads and at the same time keeps mutating the state on the main actor/thread.

Reply:
If merging and sorting lists does not end up on the main thread, it should be fine, that is a goal. If we are talking about thousands of channels per query, then we probably need to be smarter with paginated loading and unload some of them. 

I need review entity observers since these are calling `onDidChange` on the main thread. Probably a separate PR to move onDidChange to background thread (only for new code). Added a note for myself to notion.